client/manager/manager_test.go

```diff
@@ -0,0 +1,196 @@
+	"net/url"
+	"testing"
+)
+
+type testFixtures struct {
+	clientRepo client.ClientRepo
+	mgr        *ClientManager
+}
+
+var (
+	goodSecret = base64.URLEncoding.EncodeToString([]byte("my_secret"))
+)
+
+func makeTestFixtures() *testFixtures {
+	f := &testFixtures{}
+
+	dbMap := db.NewMemDB()
+
+	f.clientRepo = func() client.ClientRepo {
+		repo, err := db.NewClientRepoFromClients(dbMap, []client.Client{
+			{
+				Credentials: oidc.ClientCredentials{
+					ID:     "my_id",
+					Secret: goodSecret,
+				},
+				Metadata: oidc.ClientMetadata{
+					RedirectURIs: []url.URL{
+						{Scheme: "http", Host: "example.com", Path: "/"},
+					},
+				},
+				Admin: true,
+			},
+		})
+		if err != nil {
+			panic("Failed to create client repo: " + err.Error())
+		}
+		return repo
+	}()
+	f.mgr = NewClientManager(f.clientRepo, db.TransactionFactory(dbMap), ClientManagerOptions{})
+	return f
+}
+
+func TestMetadata(t *testing.T) {
+	tests := []struct {
+		clientID string
+		uri      string
+		wantErr  bool
+	}{
+		{
+			clientID: "my_id",
+			uri:      "http://example.com/",
+			wantErr:  false,
+		},
+	}
+
+	for i, tt := range tests {
+		f := makeTestFixtures()
+		md, err := f.mgr.Metadata(tt.clientID)
+		if err != nil {
+			t.Errorf("case %d: unexpected err: %v", i, err)
+			continue
+		}
+		if md.RedirectURIs[0].String() != tt.uri {
+			t.Errorf("case %d: manager.Metadata.RedirectURIs: want=%q got=%q", i, tt.uri, md.RedirectURIs[0].String())
+			continue
+		}
+	}
+}
+
+func TestIsDexAdmin(t *testing.T) {
+	tests := []struct {
+		clientID string
+		isAdmin  bool
+		wantErr  bool
+	}{
+		{
+			clientID: "my_id",
+			isAdmin:  true,
+			wantErr:  false,
+		},
+	}
+
+	for i, tt := range tests {
+		f := makeTestFixtures()
+		admin, err := f.mgr.IsDexAdmin(tt.clientID)
+		if err != nil {
+			t.Errorf("case %d: unexpected err: %v", i, err)
+			continue
+		}
+		if admin != tt.isAdmin {
+			t.Errorf("case %d: manager.Admin want=%t got=%t", i, tt.isAdmin, admin)
+			continue
+		}
+	}
+}
+
+func TestSetDexAdmin(t *testing.T) {
+	f := makeTestFixtures()
+	err := f.mgr.SetDexAdmin("my_id", false)
+	if err != nil {
+		t.Errorf("unexpected err: %v", err)
+	}
+	admin, _ := f.mgr.IsDexAdmin("my_id")
+	if admin {
+		t.Errorf("expected admin to be false")
+	}
+}
+
+func TestAuthenticate(t *testing.T) {
+	f := makeTestFixtures()
+	cm := oidc.ClientMetadata{
+		RedirectURIs: []url.URL{
+			url.URL{Scheme: "http", Host: "example.com", Path: "/cb"},
+		},
+	}
+
+	cc, err := f.mgr.New(client.Client{
+		Credentials: oidc.ClientCredentials{
+			ID: "baz",
+		},
+		Metadata: cm,
+	})
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	if cc.ID != "baz" {
+		t.Fatalf("Returned ClientCredentials has incorrect ID: want=baz got=%s", cc.ID)
+	}
+	ok, err := f.mgr.Authenticate(*cc)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	} else if !ok {
+		t.Fatalf("Authentication failed for good creds")
+	}
+}
+
+//func TestAuthenticate(t *testing.T) {
```

Comment:
Accidentally pushed it up, but fixed now. I was working on pulling the Authenticate method up to the manager and was changing the tests.